### PR TITLE
MM-29572: Add Bifrost support to codebase

### PR DESCRIPTION
**Pull request by agnivade**
We use a custom transport for minio when an environment variable
indicates that Mattermost is running in cloud.

The transport is used to redirect the request to the S3 endpoint set
in the config. And the scheme is set depending on if S3SSL is set.

A custom credentials provider is needed to return empty credentials
which will be overridden by the service anyways. This is just to allow
the minio client library to work transparently without knowing that
there is something else in the middle intercepting requests.

https://mattermost.atlassian.net/browse/MM-29572

### CODE REVIEW

File: services/filesstore/s3_overrides.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+// from multiple Mattermost applications, and the Mattermost service itself does not
+// have any S3 credentials.
+type customTransport struct {
+	base   http.RoundTripper
+	host   string
+	scheme string
+	cli    http.Client
+}
+
+// RoundTrip implements the transport's roundtripper implementation.
```
Comment (isacikgoz) [inline services/filesstore/s3_overrides.go:24]:
```suggestion
// RoundTrip implements the http.RoundTripper interface.
```

File: services/filesstore/s3_overrides.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+// customTransport is used to point the request to a different server.
+// This is helpful in situations where a different service is handling AWS S3 requests
+// from multiple Mattermost applications, and the Mattermost service itself does not
+// have any S3 credentials.
+type customTransport struct {
+	base   http.RoundTripper
+	host   string
+	scheme string
+	cli    http.Client
```
Comment (isacikgoz) [inline services/filesstore/s3_overrides.go:21]: Not a big fan of using `cli` for clients, it always reminds me CLI. I can be biased against it 😅 I'd use `client` instead. 0/5

File: services/filesstore/s3_overrides.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	return credentials.Value{
+		AccessKeyID:     "",
+		SecretAccessKey: "",
+		SessionToken:    "",
+		SignerType:      sign,
+	}, nil
+}
+
+// IsExpired just returns true always.
+func (cp customProvider) IsExpired() bool { return false }
```
Comment (isacikgoz) [inline services/filesstore/s3_overrides.go:59]: Does it always returns `true` or `false`?

File: services/filesstore/s3_overrides.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+// Retrieve just returns empty credentials.
+func (cp customProvider) Retrieve() (credentials.Value, error) {
+	sign := credentials.SignatureV4
+	if cp.isSignV2 {
+		sign = credentials.SignatureV2
+	}
+	return credentials.Value{
+		AccessKeyID:     "",
+		SecretAccessKey: "",
+		SessionToken:    "",
```
Comment (isacikgoz) [inline services/filesstore/s3_overrides.go:53]: Shouldn't these already be the zero value?

**Review by isacikgoz — COMMENTED** [review]
LGTM, it feels like half way done for the Bifrost with this PR :) Have some questions/suggestions though.

File: services/filesstore/s3_overrides.go (hunk 2 above)
Comment (agnivade) [reply services/filesstore/s3_overrides.go:21]: lol, yeah I have used `client` here https://github.com/mattermost/mattermost-server/pull/15948/files. Will change.

File: services/filesstore/s3_overrides.go (hunk 3 above)
Comment (agnivade) [reply services/filesstore/s3_overrides.go:59]: Always false. It doesn't really matter. It's just a dummy value.

File: services/filesstore/s3_overrides.go (hunk 4 above)
Comment (agnivade) [reply services/filesstore/s3_overrides.go:53]: Oh yeah sure.

**Comment by agnivade** [comment]
We actually cannot merge this until bifrost is running live on prod.

File: services/filesstore/s3_overrides.go (hunk 3 above)
Comment (isacikgoz) [reply services/filesstore/s3_overrides.go:56]: I see, should we update the comment then?
```suggestion
// IsExpired always returns false.
func (cp customProvider) IsExpired() bool { return false }
```

**Review by isacikgoz — APPROVED** [review]
Nicely done, a trivial thing left from my side.

File: services/filesstore/s3store.go [hunk 5]
```diff
@@ -47,7 +47,10 @@ const (
 func (b *S3FileBackend) s3New() (*s3.Client, error) {
 	var creds *credentials.Credentials
 
-	if b.accessKey == "" && b.secretKey == "" {
+	isCloud := os.Getenv("MM_CLOUD_INSTALLATION_ID") != ""
```
Comment (gabrieljackson) [inline services/filesstore/s3store.go:50]: @agnivade, we may want to have a custom env var for this behavior. I could see us continuing to have non-bifrost S3 filestores in the cloud in the future. A unique env var would be ideal for enabling or disabling bifrost in these cases.

Thanks!

Comment (agnivade) [reply services/filesstore/s3store.go:50]: Ok, that's great. This means we can push this commit without immediately breaking everything in master. And the cloud team can set that var whenever they want to enable bifrost. Let me know what name would you prefer.

Comment (gabrieljackson) [reply services/filesstore/s3store.go:50]: Yeah, exactly! Thanks!

Comment (agnivade) [reply services/filesstore/s3store.go:50]: Sounds good. Will update.

**Comment by agnivade** [comment]
@isacikgoz - One more time please. Just to be safe :)

**Review by isacikgoz — APPROVED** [review]
@agnivade no worries, looks good to me 👍

**Pull request merged**